tests/plugins/test_ddp_fully_sharded_with_full_state_dict.py

```diff
@@ -0,0 +1,202 @@
+            return
+        # resetting call_configure_sharded_model_hook attribute so that we could call
+        # configure sharded model
+        self.call_configure_sharded_model_hook = False
+        # for loading full state dict, we first need to create a new unwrapped model
+        # to load state dict and then wrapping
+        self.layer = torch.nn.Sequential(
+            torch.nn.Linear(32, 32),
+            torch.nn.ReLU(),
+            torch.nn.Linear(32, 2),
+        )
+
+    def configure_sharded_model(self) -> None:
+        for i, layer in enumerate(self.layer):
+            if i % 2 == 0:
+                self.layer[i] = wrap(layer)
+        self.layer = wrap(self.layer)
+
+    def on_load_checkpoint(self, checkpoint: Dict[str, Any]) -> None:
+        # when loading full state dict, we first need to create a new unwrapped model
+        self.setup("fit")
+
+    def configure_optimizers(self):
+        return torch.optim.SGD(self.layer.parameters(), lr=0.1)
+
+    def on_train_start(self) -> None:
+        self._assert_layer_fsdp_instance()
+
+    def on_test_start(self) -> None:
+        self._assert_layer_fsdp_instance()
+
+    def on_validation_start(self) -> None:
+        self._assert_layer_fsdp_instance()
+
+    def on_prediction_start(self) -> None:
+        self._assert_layer_fsdp_instance()
+
+    def _assert_layer_fsdp_instance(self) -> None:
+        assert isinstance(self.layer, FullyShardedDataParallel)
+        assert isinstance(self.layer.module[0], FullyShardedDataParallel)
+        assert isinstance(self.layer.module[2], FullyShardedDataParallel)
+        # root should not be resharding
+        assert self.layer.reshard_after_forward is False
+        # Assert that the nested layers are set reshard_after_forward to True
+        assert self.layer.module[0].reshard_after_forward is True
+        assert self.layer.module[2].reshard_after_forward is True
+
+
+@RunIf(min_gpus=1, skip_windows=True, fairscale_fully_sharded=True)
+def test_fully_sharded_plugin_checkpoint(tmpdir):
+    """
+    Test to ensure that checkpoint is saved correctly when using a single GPU, and all stages can be run.
+    """
+
+    model = TestFSDPModel()
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        gpus=1,
+        plugins="fsdp",
+        precision=16,
+        callbacks=[ModelCheckpoint(dirpath=tmpdir, save_last=True)],
+    )
+    _run_multiple_stages(trainer, model)
+
+
+@RunIf(min_gpus=2, skip_windows=True, fairscale_fully_sharded=True)
+def test_fully_sharded_plugin_checkpoint_multi_gpus(tmpdir):
+    """
+    Test to ensure that checkpoint is saved correctly when using multiple GPUs, and all stages can be run.
+    """
+
+    model = TestFSDPModel()
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        gpus=2,
+        plugins="fsdp",
+        precision=16,
+        callbacks=[ModelCheckpoint(dirpath=tmpdir, save_last=True)],
```

Comment:
How does this test run without the special prefix? fsdp behaves like ddp when spawning processes or not?

Reply:
good question, you mean we need `RunIf(special=True)`, right? not familiar with this special mode, @SeanNaren , do you have good answer for this.

Reply:
Wow I have no idea how this works :P adding `special=True` adds this automatically to our "special" tests, which runs each of these tests individually. It's nice for multi process tests and for tests that need to be standalone!

Reply:
somehow, really weird, for multi gpus, adding special flag, caused failure. it complains that the model checkpoint file could not be found. It is no issue for single gpu. I have to remove special flag for 2 gpus to make test pass.

Reply:
this doesn't seem correct, I should have some time to dig deeper!

Reply:
Thanks, that is indeed unexpected.

Reply:
Any progress there ?